lib/uri/generic.rb

```diff
@@ -659,7 +659,7 @@ def hostname
     # it is wrapped with brackets.
     #
     def hostname=(v)
-      v = "[#{v}]" if /\A\[.*\]\z/ !~ v && /:/ =~ v
+      v = "[#{v}]" if v&.index(':') && !v.start_with?('[') && !v.end_with?(']')
```

Comment:
`/\A\[.*\]\z/ !~ v` and `!v.start_with?('[') && !v.end_with?(']')` differ.
It should be `!(v.start_with?('[') && v.end_with?(']'))`.

Reply:
What an overlook, shame on me, this is 101. I will amend the fix. Also, I think I should start with  `start_with` in the statement because `index` call is more costly and it is more likely to not match in the future as the world progresses towards IPv6.